planner/core/point_get_plan.go

```diff
@@ -720,6 +720,14 @@ func tryWhereIn2BatchPointGet(ctx sessionctx.Context, selStmt *ast.SelectStmt) *
 			}
 			whereColNames = append(whereColNames, c.Name.Name.L)
 		}
+		// check invalid case: select * from t where (a,b) in (1,2) (#22040)
+		// if it is the case, return nil from the `TryFastPath`
+		// the stmt will be reprocess in `BuildLogicPath`
+		for _, item := range in.List {
+			if _, ok := item.(*ast.RowExpr); !ok {
+				return nil
+			}
+		}
```

Comment:
```suggestion
```

How about check length in:

![image](https://user-images.githubusercontent.com/26020263/103187485-6bf4d580-48ff-11eb-97c9-386614d71f14.png)

Reply:
Adding the checking in `func newBatchPointGetPlan` is also a solution. My initial thought was that `func tryWhereIn2BatchPointGet` should return earlier in order to avoid cost. However checking length in `func tryWhereIn2BatchPointGet` provides better understanding. I will rephrase the code.